### PR TITLE
Lay groundwork for CollateralizedSimpleInterest Adapter

### PR DESCRIPTION
**Pull request by saturnial**
This PR introduces the following types:

- `CollateralizedLoanTerms`: responsible for packing and unpacking params
- `CollateralizedAdapterErrors`: represents the errors that can be thrown by the adapter
- `CollateralizedTermsContractParameters`: the interface defining the params

This PR does not include the actual adapter or a test suite. Those will following in future PRs.

This branch merges into the main feature branch: `collateralized-simple-interest-adapter`

### CODE REVIEW

File: src/adapters/collateralized_simple_interest_loan_adapter.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
+import * as Web3 from "web3";
+import { ContractsAPI } from "src/apis";
+import { Assertions } from "src/invariants";
+import { BigNumber } from "utils/bignumber";
+import * as singleLineString from "single-line-string";
+
+const MAX_COLLATERAL_TOKEN_INDEX_HEX = TermsContractParameters.generateHexValueOfLength(39);
```
Comment (nadavhollander) [inline src/adapters/collateralized_simple_interest_loan_adapter.ts]: This should be 2, not 39 -- the max collateral index is 255

File: src/adapters/collateralized_simple_interest_loan_adapter.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+const MAX_COLLATERAL_AMOUNT_HEX = TermsContractParameters.generateHexValueOfLength(62);
```
Comment (nadavhollander) [inline src/adapters/collateralized_simple_interest_loan_adapter.ts]: This should be 23

File: src/adapters/collateralized_simple_interest_loan_adapter.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+const MAX_GRACE_PERIOD_IN_DAYS_HEX = TermsContractParameters.generateHexValueOfLength(64);
```
Comment (nadavhollander) [inline src/adapters/collateralized_simple_interest_loan_adapter.ts]: This should be 2

File: src/adapters/collateralized_simple_interest_loan_adapter.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+export interface CollateralizedTermsContractParameters {
+    collateralTokenIndex: BigNumber;
+    collateralAmount: BigNumber;
+    gracePeriodInDays: BigNumber;
+}
+
+export namespace TermsContractParameters {
```
Comment (nadavhollander) [inline src/adapters/collateralized_simple_interest_loan_adapter.ts]: Feel like this should be in a separate file that is reused among different adapters.

File: src/adapters/collateralized_simple_interest_loan_adapter.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    INVALID_TOKEN_INDEX: (tokenIndex: BigNumber) =>
+        singleLineString`Token Registry does not track a token at index
+                         ${tokenIndex.toString()}.`,
+    COLLATERAL_AMOUNT_IS_NEGATIVE: () => singleLineString`Collateral amount cannot be negative.`,
+    COLLATERAL_AMOUNT_EXCEEDS_MAXIMUM: () =>
+        singleLineString`Collateral amount exceeds maximum value of 2^92 - 1.`,
+    GRACE_PERIOD_IS_NEGATIVE: () => singleLineString`The grace period cannot be negative.`,
+    GRACE_PERIOD_EXCEEDS_MAXIMUM: () =>
+        singleLineString`The grace period exceeds the maximum value of 2^8 - 1`,
+};
```
Comment (nadavhollander) [inline src/adapters/collateralized_simple_interest_loan_adapter.ts]: Let's include invariants here for when people enter decimal values

File: src/adapters/collateralized_simple_interest_loan_adapter.ts (hunk 4 above)
Comment (saturnial) [reply src/adapters/collateralized_simple_interest_loan_adapter.ts]: That was my plan. Glad we're seeing eye-to-eye.

Comment (nadavhollander) [reply src/adapters/collateralized_simple_interest_loan_adapter.ts]: 👓

**Comment by nadavhollander** [comment]
@saturnial Is this PR still relevant now that #44 is merged?

**Comment by saturnial** [comment]
@nadavhollander wanna give this your green light? All child PRs have been merged into this one so it's acting as the main feature branch.

**Comment by nadavhollander** [comment]
Party on, wayne

**Pull request merged**